### PR TITLE
RecordsFilter should filter out accessors

### LINKED ISSUE #1215: Filter of Java 16 record accessors
### Scenario

* JaCoCo version: 0.8.7
* Operating system: unrelated
* Tool integration: unrelated
* Description of your use case: I'm using jacoco with Java 16 records.

### Current Behaviour

The record accessors are unused because I'm using them as fields. I do not want jacoco to report a missed branch for this very simple reason.

image

### Wanted Behaviour

These branches are filtered out.

### Possible Workarounds

Idk.

### PR DESCRIPTION
**Pull request by ice1000**
This PR fixes #1215, but it's not tested. I need help on authoring tests.

### CODE REVIEW

**Comment by ice1000** [comment]
```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:2.19.1:test (default-test) on project org.jacoco.core.test.validation.java16: There are test failures.
[ERROR] 
[ERROR] Please refer to /home/vsts/work/1/s/org.jacoco.core.test.validation.java16/target/surefire-reports for the individual test results.
[ERROR] -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:2.19.1:test (default-test) on project org.jacoco.core.test.validation.java16: There are test failures.

Please refer to /home/vsts/work/1/s/org.jacoco.core.test.validation.java16/target/surefire-reports for the individual test results.
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:215)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
```

Test reports??

**Comment by ice1000** [comment]
Tried to fix, let's see what happens...

**Comment by ice1000** [comment]
Nice!! All tests passed.

File: org.jacoco.core/src/org/jacoco/core/internal/analysis/filter/RecordsFilter.java
Comment (forresthopkinsa) [inline org.jacoco.core/src/org/jacoco/core/internal/analysis/filter/RecordsFilter.java:68]: Can this be a set?

Comment (ice1000) [reply org.jacoco.core/src/org/jacoco/core/internal/analysis/filter/RecordsFilter.java:68]: Of course!

Comment (marchof) [reply org.jacoco.core/src/org/jacoco/core/internal/analysis/filter/RecordsFilter.java:68]: As filters are performance critical we avoid out boxing at all. So neither List nor Set should be used for `int` values, just simple chain of comparisons and with or.

**Comment by ice1000** [comment]
@marchof I'm done, please ertake a look!

**Comment by ice1000** [comment]
The CI is failing for unknown reasons.

image

**Comment by ice1000** [comment]
```
Starting: Setup JDK
==============================================================================
Task         : Bash
Description  : Run a Bash script on macOS, Linux, or Windows
Version      : 3.214.0
Author       : Microsoft Corporation
Help         : https://docs.microsoft.com/azure/devops/pipelines/tasks/utility/bash
==============================================================================
Generating script.
========================== Starting Command Output ===========================
/bin/bash /home/vsts/work/_temp/cef2ae88-bd3a-416e-a0d6-c49f098fa19b.sh
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed

  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
  0     0    0     0    0     0      0      0 --:--:--  0:00:01 --:--:--     0
100  7073  100  7073    0     0   4724      0  0:00:01  0:00:01 --:--:--  4724

gzip: stdin: not in gzip format
tar: Child returned status 1
tar: Error is not recoverable: exiting now
##[error]Bash exited with code '2'.
Finishing: Setup JDK
```

**Comment by marchof** [comment]
@ice1000 Our CI (and the internet in general) is a bit flaky. Sometime JDK downloads do fail. As you have green builds for most of the JDKs don't worry for now. With your next commit it will probably work again.

**Comment by ice1000** [comment]
> @ice1000 Our CI (and the internet in general) is a bit flaky. Sometime JDK downloads do fail. As you have green builds for most of the JDKs don't worry for now. With your next commit it will probably work again.

Do you expect a next commit? Is there anything else I can do to get this merged? 🔮

**Comment by marchof** [comment]
@ice1000 We need more commits to finalize this: Can you please add unit test to `RecordsFilterTest`. You find lots of unit tests in the same package that you can use as reference.

**Comment by ice1000** [comment]
@marchof what do you think about the new two tests I've added? Do I need more?

File: org.jacoco.core.test/src/org/jacoco/core/internal/analysis/filter/RecordsFilterTest.java
Comment (marchof) [inline org.jacoco.core.test/src/org/jacoco/core/internal/analysis/filter/RecordsFilterTest.java:147]: Why should the getter *not* be filtered here? Looks like the test setup is wrong. I would recommend to start with postive test cases like `should_filter_int_accessor` etc.

Afterwards add negative tests e.g. `should_not_filter_accessor_when_name_does_not_match`

Comment (ice1000) [reply org.jacoco.core.test/src/org/jacoco/core/internal/analysis/filter/RecordsFilterTest.java:147]: Ohhh you're right, I got the method name wrong. It's actually a "should filter" test. I'll correct

Comment (marchof) [reply org.jacoco.core.test/src/org/jacoco/core/internal/analysis/filter/RecordsFilterTest.java:147]: In this case you should call assertMethodIgnored(m);

Comment (ice1000) [reply org.jacoco.core.test/src/org/jacoco/core/internal/analysis/filter/RecordsFilterTest.java]: Fixed!

Comment (ice1000) [reply org.jacoco.core.test/src/org/jacoco/core/internal/analysis/filter/RecordsFilterTest.java]: Fixed!

**Comment by ice1000** [comment]
@marchof done! I'm adding one more "should_not" test....

**Comment by ice1000** [comment]
Completed!

**Comment by marchof** [comment]
Nice! If I run `RecordsFilterTest` with code code coverage I see some missed cases. At JaCoCo we aim 100% functional test coverage for our own code base. Can you please add tests for the remaining return types? Also ot looks like we need a negative test when `GETFIELD` is not directly followed by any kind of return statement.

image

**Comment by ice1000** [comment]
@marchof before I finish this, do you have an ETA for the next release of jacoco?

**Comment by marchof** [comment]
We don't commit to any release dates, see FAQ. But latest with Java 20 in March there will be a new JaCoCo release. So let's finish this so it can be part of it.

**Comment by ice1000** [comment]
> We don't commit to any release dates, see FAQ. But latest with Java 20 in March there will be a new JaCoCo release. So let's finish this so it can be part of it.

Cool!!!!!

Waiting for CI....

File: org.jacoco.core/src/org/jacoco/core/internal/analysis/filter/RecordsFilter.java
Comment (marchof) [inline org.jacoco.core/src/org/jacoco/core/internal/analysis/filter/RecordsFilter.java:89]: We always use braces for blocks. Can you please add them here and above?

```suggestion
			if (cursor == null) {
				return false;
			}
```

Comment (ice1000) [reply org.jacoco.core/src/org/jacoco/core/internal/analysis/filter/RecordsFilter.java]: Fixed

File: org.jacoco.core/src/org/jacoco/core/internal/analysis/filter/RecordsFilter.java
Comment (marchof) [inline org.jacoco.core/src/org/jacoco/core/internal/analysis/filter/RecordsFilter.java:86]: Same here:

```suggestion
			if (!(cursor instanceof FieldInsnNode)) {
				return false;
			}
			if (!((FieldInsnNode) cursor).name.equals(m.name)) {
				return false;
			}
```

Comment (ice1000) [reply org.jacoco.core/src/org/jacoco/core/internal/analysis/filter/RecordsFilter.java]: Fixed

**Comment by marchof** [comment]
Alternatively please add the following block to changes.html (as the last point under New Features:

```
  <li>Component accessors generated by the Java compilers for records are filtered
      out during generation of report. Contributed by Tesla Zhang‮
      (GitHub <a href="https://github.com/jacoco/jacoco/issues/1393">#1393</a>).</li>
```

**Comment by ice1000** [comment]
I don't see the button in the tutorial. I invited you as a collaborator: https://github.com/ice1k/jacoco/invitations

**Comment by marchof** [comment]
> I don't see the button in the tutorial. I invited you as a collaborator: https://github.com/ice1k/jacoco/invitations

Probably this is because the repository belongs to a different user (ice1000 vs ice1k). As a collaborator I was now able to push.

**Comment by ice1000** [comment]
Merge?

**Comment by marchof** [comment]
@Godin Can you please have a look?

**Comment by marchof** [comment]
> Merge?

Not yet, a second committer (@Godin) will now do the final review.

File: org.jacoco.core/src/org/jacoco/core/internal/analysis/filter/RecordsFilter.java
Comment (Godin) [inline org.jacoco.core/src/org/jacoco/core/internal/analysis/filter/RecordsFilter.java:80]: Probably a matter of taste, but IMO will be better to move comment into body of the `if`

```suggestion
			if (!m.desc.startsWith("()")) {
				// method with parameters
				return false;
			}
```
or remove it entirely.

File: org.jacoco.doc/docroot/doc/changes.html
Comment (Godin) [inline org.jacoco.doc/docroot/doc/changes.html:34]: On this line GitHub shows the following warning

> This file contains bidirectional Unicode text that may be interpreted or compiled differently than what appears below. To review, open the file in an editor that reveals hidden Unicode characters. [Learn more about bidirectional Unicode characters](https://github.co/hiddenchars)

because there is `U+202E` at the end of the line.

And this file in Chrome looks as following

Screenshot 2023-02-05 at 01 38 07

So I believe that this character should be removed.

**Review by Godin — CHANGES_REQUESTED** [review]

File: org.jacoco.core/src/org/jacoco/core/internal/analysis/filter/RecordsFilter.java
Comment (ice1000) [reply org.jacoco.core/src/org/jacoco/core/internal/analysis/filter/RecordsFilter.java:80]: Done

File: org.jacoco.doc/docroot/doc/changes.html
Comment (ice1000) [reply org.jacoco.doc/docroot/doc/changes.html:34]: Done

**Comment by ice1000** [comment]
@Godin all done, please retake a look!

File: org.jacoco.doc/docroot/doc/changes.html
Comment (marchof) [reply org.jacoco.doc/docroot/doc/changes.html]: My bad. Thanks for spotting and fixing it!

**Comment by ice1000** [comment]
ping 👻

**Comment by Godin** [comment]
@ice1000 We are working on this project in our spare time when we have it. We will appreciate it if you'll let us manage our time by ourselves - no need to send pings.

**Review by Godin — APPROVED** [review]

**Comment by forresthopkinsa** [comment]
You love to see it. Thanks for the work folks

**Pull request merged**
